src/mixins/record.js

```diff
@@ -0,0 +1,178 @@
+        .then(() => this.validateRecord())
+        .then(() => {
+          if (isNew) {
+            return this.$ComposeAPI.recordCreate(this.record)
+          } else {
+            return this.$ComposeAPI.recordUpdate(this.record)
+          }
+        })
+        .catch(err => {
+          const { details = undefined } = err
+          if (!!details && Array.isArray(details) && details.length > 0) {
+            this.errors.push(...details)
+            throw new Error(this.$t('notification.record.validationErrors'))
+          }
+
+          throw err
+        })
+        .then((record) => this.record.apply(record))
+        .then(() => this.dispatchUiEvent('afterFormSubmit'))
+        .then(() => {
+          this.$router.push({ name: route, params: { ...this.$route.params, recordID: this.record.recordID } })
+        })
+        .catch(this.defaultErrorHandler(this.$t(
+          isNew
+            ? 'notification.record.createFailed'
+            : 'notification.record.updateFailed',
+        )))
+    },
+
+    /**
+     * On delete, preserve user's view. Show a notification that the record
+     * has been deleted.
+     */
+    handleDelete () {
+      return this
+        .dispatchUiEvent('beforeDelete')
+        .then(() => this.$ComposeAPI.recordDelete(this.record))
+        .then(() => {
+          this.record.deletedAt = (new Date()).toISOString()
+        })
+        .then(() => this.dispatchUiEvent('afterDelete'))
+        .catch(this.defaultErrorHandler(this.$t('notification.record.deleteFailed')))
+    },
+
+    /**
+     * Validates record and dispatches onFormSubmitError
+     *
+     * onFormSubmitError is dispatched only if there are record errors,
+     * if not, we continue with form submit handling
+     *
+     * After onFormSubmitError, record is re-validated and if errors
+     * are still present, we stop form submit handing
+     *
+     * @returns {Promise<void>}
+     */
+    async validateRecord () {
+      this.errors = this.validator.run(this.record)
+      if (this.errors.valid()) {
+        return
+      }
+
+      await this.dispatchUiEvent('onFormSubmitError')
+
+      this.errors = this.validator.run(this.record)
+      if (!this.errors.valid()) {
+        throw new Error(this.$t('notification.record.validationErrors'))
+      }
+    },
+
+    /**
+     * Returns errors, filtered for a specific field
+     * @param name
+     * @returns {validator.Validated} filtered validation results
+     */
+    fieldErrors (name) {
+      if (!this.errors) {
+        return new validator.Validated()
+      }
+
+      return this.errors.filterByMeta('field', name)
+    },
+
+    /**
+     * Generic event dispatcher for ui:compose:record-page resource type
+     *
+     * This is used when deleting, updating, creating
+     * records and where validation errors occur
+     *
+     * @param eventType
+     */
+    dispatchUiEvent (eventType) {
+      const resourceType = 'ui:compose:record-page'
```

Comment:
make this configurable:
 - `ui:compose:record-page`
 - `ui:compose:admin-record-page`

maybe it's best if you call `getUiEventResourceType()` from component (and fallback to `ui:compose:record-page`)